src/test/scala/org/broadinstitute/dsde/firecloud/service/NihApiServiceSpec.scala

```diff
@@ -0,0 +1,96 @@
+import org.broadinstitute.dsde.firecloud.utils.DateUtils
+import spray.http.StatusCodes._
+
+import scala.concurrent.ExecutionContext
+
+class NihApiServiceSpec extends ApiServiceSpec {
+  val dbGapAuthorizedUsersGroupName = FireCloudConfig.Nih.rawlsGroupName
+  val validJwtOnWhitelist = JWTWrapper("eyJhbGciOiJIUzI1NiJ9.ZmlyZWNsb3VkLWRldg.NPXbSpTmAOUvJ1HX85TauAARnlMKfqBsPjumCC7zE7s")
+  val validJwtNotOnWhitelist = JWTWrapper("eyJhbGciOiJIUzI1NiJ9.bm90LWEtcmVhbC11c2Vy.L6JKlzZ9zp2mvDAbhGKRJIbW99x4cN-UjWJNZuSdnbc")
+
+  case class TestApiService(agoraDao: MockAgoraDAO, rawlsDao: MockRawlsDAO, searchDao: MockSearchDAO, thurloeDao: MockThurloeDAO, googleDao: MockGoogleServicesDAO)(implicit val executionContext: ExecutionContext) extends ApiServices
+
+  def withDefaultApiServices[T](testCode: TestApiService => T): T = {
+    val apiService = new TestApiService(new MockAgoraDAO, new MockRawlsDAO, new MockSearchDAO, new MockThurloeDAO, new MockGoogleServicesDAO)
+    testCode(apiService)
+  }
+
+  "NihApiService" should "return NotFound when GET-ting a profile with no NIH username" in withDefaultApiServices { services =>
+    Get("/nih/status") ~> dummyUserIdHeaders(services.thurloeDao.normalUserId) ~> sealRoute(services.nihRoutes) ~> check {
+      status should equal(NotFound)
+    }
+  }
+
+  it should "return NotFound when GET-ting a non-existent profile" in withDefaultApiServices { services =>
+    Get("/nih/status") ~> dummyUserIdHeaders("userThatDoesntExist") ~> sealRoute(services.nihRoutes) ~> check {
+      status should equal(NotFound)
+    }
+  }
+
+  it should "return BadRequest when NIH linking with an invalid JWT" in withDefaultApiServices { services =>
+    Post("/nih/callback", JWTWrapper("bad-token")) ~> dummyUserIdHeaders(services.thurloeDao.normalUserId) ~> sealRoute(services.nihRoutes) ~> check {
+      status should equal(BadRequest)
+    }
+    assert(!services.rawlsDao.groups(dbGapAuthorizedUsersGroupName).contains(services.thurloeDao.normalUserId))
+  }
+
+  it should "return OK when NIH linking with a valid JWT and the user is on the whitelist" in withDefaultApiServices { services =>
+    assert(!services.rawlsDao.groups(dbGapAuthorizedUsersGroupName).contains(services.thurloeDao.normalUserId))
+    Post("/nih/callback", validJwtOnWhitelist) ~> dummyUserIdHeaders(services.thurloeDao.normalUserId) ~> sealRoute(services.nihRoutes) ~> check {
+      println(response)
+      status should equal(OK)
+    }
+    assert(services.rawlsDao.groups(dbGapAuthorizedUsersGroupName).contains(services.thurloeDao.normalUserId))
+  }
+
+  it should "return InternalServerError when NIH linking with a valid JWT and the user is NOT on the whitelist" in withDefaultApiServices { services =>
+    Post("/nih/callback", validJwtNotOnWhitelist) ~> dummyUserIdHeaders(services.thurloeDao.normalUserId) ~> sealRoute(services.nihRoutes) ~> check {
+      status should equal(InternalServerError)
```

Comment:
the internalservererror was done as part of the quick hotfix to prevent non-whitelisted NIH users from getting access to controlled TCGA data. we do have a ticket to improve the error messaging and everything related to this situation